### PR TITLE
SUBS-1825-release-date

### PR DESCRIPTION
**Pull request by haseeb-gh**
- added support for study and sample release dates.
- updated implementation to send multiple submissions when addind new data objects with release dates.
- added/updated unit tests.

### CODE REVIEW

File: src/main/java/uk/ac/ebi/subs/ena/processor/ENAProcessor.java [hunk 1]
```diff
@@ -169,5 +190,89 @@ private String centerName(SubmissionEnvelope submissionEnvelope) {
         return paramMap;
     }
 
+    private Study setDefaultStudyReleaseDate(SubmissionEnvelope submissionEnvelope, Study study) {
+        Date subDate = submissionEnvelope.getSubmission().getSubmissionDate();
+
+        if (subDate != null && !study.isAccessioned() && study.getReleaseDate() == null) {
```
Comment (ke4) [inline src/main/java/uk/ac/ebi/subs/ena/processor/ENAProcessor.java:196]: What happens if `subdate` is null or study is accessioned?

File: src/main/java/uk/ac/ebi/subs/ena/processor/ENAProcessor.java [hunk 2]
```diff
@@ -169,5 +190,89 @@ private String centerName(SubmissionEnvelope submissionEnvelope) {
...
+            study.setReleaseDate(convertSubmissionDate(submissionEnvelope.getSubmission().getSubmissionDate()));
+        }
+
+        return study;
+    }
+
+    private Sample setDefaultSampleReleaseDate(SubmissionEnvelope submissionEnvelope, Sample sample) {
+        Date subDate = submissionEnvelope.getSubmission().getSubmissionDate();
```
Comment (ke4) [inline src/main/java/uk/ac/ebi/subs/ena/processor/ENAProcessor.java:204]: The same question: What happens if `subdate` is null or sample is accessioned?

File: src/main/java/uk/ac/ebi/subs/ena/processor/ENAProcessor.java [hunk 3]
```diff
@@ -169,5 +190,89 @@ private String centerName(SubmissionEnvelope submissionEnvelope) {
...
+
+            Class<? extends ActionService> service = serviceSubmittableEntry.getKey();
+            Object serviceSubmittable = serviceSubmittableEntry.getValue();
+
+            groupByReleaseDate(serviceSubmittable).entrySet().stream().forEach(dateSubmittableEntry -> {
+
+                Optional<LocalDate> date = dateSubmittableEntry.getKey();
+                Object dateSubmittable = dateSubmittableEntry.getValue();
+
+                Map<Class<? extends ActionService>, Object> group = groups.get(date);
```
Comment (ke4) [inline src/main/java/uk/ac/ebi/subs/ena/processor/ENAProcessor.java:239]: I am not sure why don't you simple use `groups.put(date, Map.of(service, dateSubmittable))`
From API doc `LinkedhashMap` - `put`method: Associates the specified value with the specified key in this map. If the map previously contained a mapping for the key, the old value is replaced.

```suggestion
                groups.put(date, Map.of(service, dateSubmittable));
```

File: src/test/java/uk/ac/ebi/subs/ena/submission/FullSubmissionServiceTest.java [hunk 4]
```diff
@@ -44,7 +47,7 @@
     @Value("${ena.password}")
     String ftpPassword;
 
-    public static final int SUBMITTABLE_COUNT = 10;
+    public static final int SUBMITTABLE_COUNT = 1;
```
Comment (ke4) [inline src/test/java/uk/ac/ebi/subs/ena/submission/FullSubmissionServiceTest.java]: Why did you change it to 1?

**Review by ke4 — CHANGES_REQUESTED** [review]
Looked really good 👍 , but I have some comments and change requests related to the changes.
Cheers,
Karel

File: src/main/java/uk/ac/ebi/subs/ena/processor/ENAProcessor.java (hunk 1 above)
Comment (haseeb-gh) [reply src/main/java/uk/ac/ebi/subs/ena/processor/ENAProcessor.java:196]: this case is special because an accessioned study or sample means they have already been created and are being requested for an update. so here, we dont use submission date as its quite possible that the user who is requesting an update does not want to change release date. if they do want want to update it then they should specify the release date in the submittable object instead.

File: src/main/java/uk/ac/ebi/subs/ena/processor/ENAProcessor.java (hunk 3 above)
Comment (haseeb-gh) [reply src/main/java/uk/ac/ebi/subs/ena/processor/ENAProcessor.java:239]: if i got this right then what you're suggesting will replace an existing group from the groups map. which is not what we want. Instead, if a group is present already then we take it and update it.

please let me know if i got it wrong?

File: src/test/java/uk/ac/ebi/subs/ena/submission/FullSubmissionServiceTest.java (hunk 4 above)
Comment (haseeb-gh) [reply src/test/java/uk/ac/ebi/subs/ena/submission/FullSubmissionServiceTest.java]: thanks for catching this. i temporarily changed to generate some xmls as an example for ENA. i ll set it back to 10.

File: src/main/java/uk/ac/ebi/subs/ena/processor/ENAProcessor.java [hunk 5]
```diff
@@ -169,5 +190,89 @@ private String centerName(SubmissionEnvelope submissionEnvelope) {
         return paramMap;
     }
 
+    private Study setDefaultStudyReleaseDate(SubmissionEnvelope submissionEnvelope, Study study) {
+        Date subDate = submissionEnvelope.getSubmission().getSubmissionDate();
+
+        if (subDate != null && !study.isAccessioned() && study.getReleaseDate() == null) {
+            study.setReleaseDate(convertSubmissionDate(submissionEnvelope.getSubmission().getSubmissionDate()));
```
Comment (ke4) [inline src/main/java/uk/ac/ebi/subs/ena/processor/ENAProcessor.java]: You could use `subDate` variable instead of getting it again from the envelop

```suggestion
            study.setReleaseDate(convertSubmissionDate(subDate));
```

File: src/main/java/uk/ac/ebi/subs/ena/processor/ENAProcessor.java (hunk 1 above)
Comment (ke4) [reply src/main/java/uk/ac/ebi/subs/ena/processor/ENAProcessor.java:196]: OK, got it

File: src/main/java/uk/ac/ebi/subs/ena/processor/ENAProcessor.java (hunk 3 above)
Comment (ke4) [reply src/main/java/uk/ac/ebi/subs/ena/processor/ENAProcessor.java:239]: I think with this line you change the reference anyway: `group.put(service, dateSubmittable);`. Am I wrong?

File: src/main/java/uk/ac/ebi/subs/ena/processor/ENAProcessor.java (hunk 5 above)
Comment (haseeb-gh) [reply src/main/java/uk/ac/ebi/subs/ena/processor/ENAProcessor.java]: good catch. will fix this as well.

Comment (ke4) [reply src/main/java/uk/ac/ebi/subs/ena/processor/ENAProcessor.java]: just use the suggestion from github :-)

File: src/main/java/uk/ac/ebi/subs/ena/processor/ENAProcessor.java (hunk 3 above)
Comment (haseeb-gh) [reply src/main/java/uk/ac/ebi/subs/ena/processor/ENAProcessor.java:239]: that line updates the group. i think your suggestion was about line 239. that would have always replaced an existing group.

Comment (ke4) [reply src/main/java/uk/ac/ebi/subs/ena/processor/ENAProcessor.java]: Ok, I got it

But what about this:

```suggestion
                Map<Class<? extends ActionService>, Object> group = groups.get(date);
                groups.put(date, group == null ? Map.of(service, dateSubmittable) : group.put(service, dateSubmittable));
```

Comment (ke4) [reply src/main/java/uk/ac/ebi/subs/ena/processor/ENAProcessor.java:239]: Ok, got it. I was wrong

**Review by ke4 — APPROVED** [review]
